ChopChop/Parser/RecipeParser.swift

```diff
@@ -0,0 +1,224 @@
+//
+//  Parser.swift
+//  ChopChop
+//
+//  Created by Cao Wenjie on 14/3/21.
+//
+import Foundation
+
+class RecipeParser {
+    
+    // matches 1. or 1)
+    static let stepsIndexRegex = "[1-9][0-9]*(\\.|\\))\\s"
+    // matches 2 or 2.5
+    static let intOrDecimal = "[1-9]\\d*(\\.\\d+)?"
+    // matches 2
+    static let integer = "[1-9]\\d*"
+    // matches 1/2 or 1 / 2
+    static let fraction = "[1-9]\\s*/\\s*[1-9]"
+    static let whitespace = "\\s+"
+    static let optionalWhiteSpace = "\\s*"
+    // places all units into a group like (kg|g|cup)
+    static let units: String = (Array(UnitsMapping.volumeWordMap.keys) + Array(UnitsMapping.massWordMap.keys))
+        .joined(separator: "|")
+    
+    /**
+     Parses a chunk of instructions into an array of steps. Parsing is done differently depending on
+     whether the instructions are already numbered.
+     */
+    static func parseInstructions(instructions: String) -> [String] {
+        let regex = NSRegularExpression(stepsIndexRegex)
+
+        let indexes = regex.matches(in: instructions, options: [],
+                                    range: NSRange(location: 0, length: instructions.utf16.count))
+        // when instructions are numbered, just parse base on the numbering
+        if !indexes.isEmpty {
+            var steps = [String]()
+            for idx in 1..<indexes.count {
+                let startIndex = instructions.index(instructions.startIndex,
+                                                    offsetBy: indexes[idx - 1].range.upperBound)
+                let endIndex = instructions.index(instructions.startIndex,
+                                                  offsetBy: indexes[idx].range.lowerBound - 1)
+                steps.append(String(instructions[startIndex...endIndex])
+                                .trimmingCharacters(in: .whitespacesAndNewlines))
+            }
+            let lastStart = instructions.index(instructions.startIndex,
+                                               offsetBy: indexes[indexes.count - 1].range.upperBound)
+            steps.append(String(instructions[lastStart..<instructions.endIndex])
+                            .trimmingCharacters(in: .whitespacesAndNewlines))
+            return steps
+        // when instructions are not numbered, each sentence is taken as a step
+        } else {
+            return instructions.components(separatedBy: ".")
+                .dropLast()
+                .map({ $0 + "." })
```

Comment:
`components(separatedBy:)` is a `NSString` method, I believe the one we should be using is [`split(separator:)`](https://developer.apple.com/documentation/swift/string/2894564-split)

Reply:
yup i will change this.